### DISCUSSION TITLE
Use extended ansi colors in format object?

### DISCUSSION THREAD

**Discussion by chtenb** [wezterm/wezterm#5355] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

newest

### Ask your question!

Several wezterm configuration functions accept `FormatItem` objects https://wezfurlong.org/wezterm/config/lua/wezterm/format.html. Using these objects one can specify attributes and colors. To choose ansi colors 0-15, the constructor `AnsiColor` is used. How would one specify extended ansi colors `16-255`?

**Comment by wez** [comment]
There no direct support, but you can use the underlying ANSI escape sequences for the extended palette indices in most cases.

https://wezfurlong.org/wezterm/escape-sequences.html#csi-385-foreground-color-palette-index

**Reply by chtenb** [reply]
Thanks! I tried to play around with this, but I noticed that these escape codes would not respect the theme colors. Is that intended behavior? (I'm calling them through `window:set_right_status(wez.format(...`)

**Reply by wez** [reply]
What are you trying exactly?

**Reply by chtenb** [reply]
I'm trying to format the status bar to use colors I defined in my color theme. I've defined a color palette using the ansi color numbers, which I'm referencing in all my terminal apps themes. I would like to reference these colors in the wezterm config too. I expected that using escape codes in wez.format would respect the colors from my color theme, but it seems they refer to the builtin colors instead? For a (work-in-progress) example of a theme I use, see https://github.com/chtenb/neo-ansi/blob/ed528efe610a49da3e5ef6e321413f79ae1ea5fd/terms/wezterm/generated/lux-neo-ansi.toml

**Reply by wez** [reply]
Can you show me how you are using `wezterm.format` and what you're seeing vs. what you're expecting?

**Reply by chtenb** [reply]
Example: `window:set_right_status(wez.format({{ Text = '\x1b[48:5:232:0mhello' }}))`, where 232 is a color I defined as `#f8f1e0`. This renders like: 
![image](https://github.com/wez/wezterm/assets/1030961/ee1d30af-605c-43c3-bede-ba3577b4701b)
Notice the background is black, which corresponds to the default for color for 232. I had expected to see `#f8f1e0`.

**Reply by chtenb** [reply]
I did some more investigating, and noticed that this behavior is also happening when using the `AnsiColor` attribute in the FormatItem.

**Reply by wez** [reply]
This is what I see:

![image](https://github.com/wez/wezterm/assets/117777/ea293d15-7d4c-42de-b9c6-f9518e1f5921)

Both using the raw escape sequence (without `wezterm.format`), and embedding it into the `Text = ...` syntax work for me here; both the left and right status areas are picking up the yellow background from this minimal config:

```lua
local wezterm = require 'wezterm'
local config = wezterm.config_builder()

config.colors = {
  indexed = {
    [232] = 'yellow',
  }
}

wezterm.on('update-status', function(window, pane)
  window:set_right_status("\x1b[48:5:232:0mhello")
  window:set_left_status(wezterm.format({{ Text = '\x1b[48:5:232:0mhello' }}))
end)

return config
```

I suspect that you are encountering a different issue. Perhaps your color scheme isn't being picked up?

**Reply by chtenb** [reply]
Huh, that's interesting. I can confirm that it works as expected when I define the colors in the config table directly, like you did in your minimal config above. But not when the colors are defined in a colorscheme file (toml).

**Reply by chtenb** [reply]
I'm pretty sure my color scheme is picked up. When I change the background color, the color changes (after reloading config). Moreover, I'm referencing the ansi and indexed colors in colorschemes in other terminal apps, and they do work there like I expect.

**Reply by wez** [reply]
are you using set_config_overrides to update the colors, or is the color scheme part of the overall config?

**Reply by chtenb** [reply]
Yes, that's correct. I use `set_config_overrides ` in the `window-config-reloaded` event to dynamically switch between light and dark mode, based on the system appearance.

**Reply by chtenb** [reply]
By your inquiry I assume that this behavior is not intended. Should I create an issue for this?

**Reply by wez** [reply]
It sounds weird enough that I think it is worth the effort to dig deeper, so, please file with the minimal config that reproduces this!

**Reply by chtenb** [reply]
For future reference, I created this issue: https://github.com/wez/wezterm/issues/5478

**Comment by nekowinston** [comment]
If you have access to `window:effective_config()`, you can use `config.resolved_palette.indexed[n]`.
I'm using it to format my tab bar here:
https://github.com/nekowinston/wezterm-bar/blob/e96b81460b3ad11a7461934dcb7889ce5079f97f/plugin/init.lua#L204

**Reply by chtenb** [reply]
Thanks for the reply. I think I have the window object available in the current case yes, since I'm writing eventhandlers. Not sure if that will be true in all situations.
I notice that this method only returns the colors when they are actually overridden in the palette, otherwise it returns nil. Specifically, the numbers 0-15 don't work, because they are defined differently in the theme file.
